### PR TITLE
Make InMemory object store track last modified time for each entry

### LINKED ISSUE #3782: Make `InMemory` object store track last modified time for each entry
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The `InMemory` object stores currently [spoofs](https://github.com/apache/arrow-rs/blob/master/object_store/src/memory.rs#L148) the `last_modified` field of the fetched `ObjectMeta`s. In some cases, it is important for this value to be correct. 

One such case is [deducing the `DeltaTable` version](https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L737-L739) based on a user-provided timestamp in time travel querying—if the `last_modified` field is wrong the wrong version of the table will be loaded. Admittedly, no one should be using an in-memory store for actual work, but it would still be very handy to have this working for testing.

**Describe the solution you'd like**
Keep a separate map of `last_modified` timestamps, and update it on each write call (`put`, `copy`, etc.).

**Describe alternatives you've considered**
Don't use `InMem` store whenever `last_modified` field is crucial for the logic.

**Additional context**
This could be a good first issue.

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #3782

# Rationale for this change
 

Explain in #3782

# What changes are included in this PR?

In `struct InMemoryUpload`, it can store btye with timestamp

# Are there any user-facing changes?

### CODE REVIEW

**Review by gruuya — COMMENTED** [review]
I'd argue that the `copy`/`copy_if_not_exists` methods should also alter the last modified timestamps if we're going to emulate local FS behavior:
```shell
$ touch test
$ date -r test
Fri Mar  3 09:32:25 CET 2023
$ cp test test_cp
$ date -r test_cp
Fri Mar  3 09:33:50 CET 2023
```

That said, `rename` also [uses](https://github.com/apache/arrow-rs/blob/master/object_store/src/lib.rs#L399-L400) `copy` (and likewise for `rename_if_not_exists`), but it probably shouldn't affect the timestamp:
```shell
$ mv test test_mv
$ date -r test_mv
Fri Mar  3 09:32:25 CET 2023
```

File: object_store/src/memory.rs [hunk 1]
```diff
@@ -273,22 +278,23 @@ impl InMemory {
         }
     }
 
-    async fn get_bytes(&self, location: &Path) -> Result<Bytes> {
+    async fn get_bytes(&self, location: &Path) -> Result<(Bytes, DateTime<Utc>)> {
```
Comment (tustvold) [inline object_store/src/memory.rs:281]:
```suggestion
    async fn entry(&self, location: &Path) -> Result<(Bytes, DateTime<Utc>)> {
```

File: object_store/src/memory.rs [hunk 2]
```diff
@@ -33,6 +33,8 @@ use std::sync::Arc;
 use std::task::Poll;
 use tokio::io::AsyncWrite;
 
+type StorageType = Arc<RwLock<BTreeMap<Path, (Bytes, DateTime<Utc>)>>>;
```
Comment (tustvold) [inline object_store/src/memory.rs:36]:
```suggestion
type Entry = (Bytes, DateTime<Utc>);
type StorageType = Arc<RwLock<BTreeMap<Path, Entry>>>;
```

File: object_store/src/memory.rs [hunk 3]
```diff
@@ -150,7 +155,7 @@ impl ObjectStore for InMemory {
         Ok(ObjectMeta {
             location: location.clone(),
             last_modified,
```
Comment (tustvold) [inline object_store/src/memory.rs:157]: I think this should return the timestamp from the entry?

File: object_store/src/memory.rs [hunk 4]
```diff
@@ -181,7 +186,7 @@ impl ObjectStore for InMemory {
                 Ok(ObjectMeta {
                     location: key.clone(),
                     last_modified,
```
Comment (tustvold) [inline object_store/src/memory.rs:188]: This should be the value from the entry?

File: object_store/src/memory.rs [hunk 5]
```diff
@@ -225,7 +230,7 @@ impl ObjectStore for InMemory {
                 let object = ObjectMeta {
                     location: k.clone(),
                     last_modified,
```
Comment (tustvold) [inline object_store/src/memory.rs:232]: Ditto here

**Review by tustvold — COMMENTED** [review]
I think we should be returning this timestamp as the last_modified when we return ObjectMeta, and possibly verify that we are setting it consistently in all the mutating methods (such as copy)

File: object_store/src/memory.rs [hunk 6]
```diff
@@ -273,22 +275,23 @@ impl InMemory {
         }
     }
 
-    async fn get_bytes(&self, location: &Path) -> Result<Bytes> {
+    async fn enty(&self, location: &Path) -> Result<(Bytes, DateTime<Utc>)> {
```
Comment (tustvold) [inline object_store/src/memory.rs:278]:
```suggestion
    async fn entry(&self, location: &Path) -> Result<(Bytes, DateTime<Utc>)> {
```

**Review by tustvold — APPROVED** [review]

File: object_store/src/memory.rs [hunk 7]
```diff
@@ -238,13 +240,13 @@ impl ObjectStore for InMemory {
     }
 
     async fn copy(&self, from: &Path, to: &Path) -> Result<()> {
-        let data = self.get_bytes(from).await?;
+        let data = self.entry(from).await?;
```
Comment (tustvold) [inline object_store/src/memory.rs:243]: I think this should probably generate a new last modified timestamp

File: object_store/src/memory.rs [hunk 8]
```diff
@@ -238,13 +240,13 @@ impl ObjectStore for InMemory {
...
         self.storage.write().insert(to.clone(), data);
         Ok(())
     }
 
     async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> Result<()> {
-        let data = self.get_bytes(from).await?;
+        let data = self.entry(from).await?;
```
Comment (tustvold) [inline object_store/src/memory.rs:249]: Ditto here

File: object_store/src/memory.rs (hunk 7 above)
Comment (Weijun-H) [reply object_store/src/memory.rs:243]: like this `self.storage.write().insert(to.clone(), (data.0, DateTime::from(Utc::now())));`?

**Comment by tustvold** [comment]
Thanks again

**Pull request merged**